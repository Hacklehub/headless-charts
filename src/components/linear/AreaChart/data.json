[
  {
    "year": 2012,
    "macbook": 23.22,
    "services": 12.89,
    "wearables": 10.76,
    "ipad": 30.94,
    "iphone": 78.69
  },
  {
    "year": 2013,
    "macbook": 21.48,
    "services": 16.05,
    "wearables": 10.12,
    "ipad": 31.98,
    "iphone": 91.28
  },
  {
    "year": 2014,
    "macbook": 24.08,
    "services": 18.06,
    "wearables": 8.379,
    "ipad": 30.28,
    "iphone": 101.99
  },
  {
    "year": 2015,
    "macbook": 25.47,
    "services": 19.91,
    "wearables": 10.07,
    "ipad": 23.23,
    "iphone": 155.04
  },
  {
    "year": 2016,
    "macbook": 22.83,
    "services": 24.35,
    "wearables": 11.13,
    "ipad": 20.63,
    "iphone": 136.7
  },
  {
    "year": 2017,
    "macbook": 25.57,
    "services": 32.7,
    "wearables": 12.83,
    "ipad": 18.8,
    "iphone": 139.34
  },

  {
    "year": 2018,
    "macbook": 25.2,
    "services": 39.75,
    "wearables": 17.38,
    "ipad": 18.38,
    "iphone": 164.89
  },
  {
    "year": 2019,
    "macbook": 25.74,
    "services": 46.29,
    "wearables": 24.48,
    "ipad": 21.28,
    "iphone": 142.38
  },
  {
    "year": 2020,
    "macbook": 28.62,
    "services": 53.77,
    "wearables": 30.62,
    "ipad": 23.72,
    "iphone": 137.78
  },

  {
    "year": 2021,
    "macbook": 35.19,
    "services": 68.42,
    "wearables": 38.37,
    "ipad": 31.86,
    "iphone": 191.97
  },
  {
    "year": 2022,
    "macbook": 40.18,
    "services": 78.13,
    "wearables": 41.24,
    "ipad": 29.29,
    "iphone": 205.49
  }
]
