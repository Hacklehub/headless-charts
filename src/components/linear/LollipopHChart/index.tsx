import React, { useEffect } from 'react';
import { axisBottom, axisLeft, axisRight, axisTop } from 'd3-axis';
import { defaultChartClassNames } from '../../../utils';
import { twMerge } from 'tailwind-merge';
import { max, min } from 'd3-array';
import { pointer, select, selectAll } from 'd3-selection';
import { scaleLinear, scalePoint } from 'd3-scale';
import {
  symbol,
  symbolCircle,
  symbolCross,
  symbolDiamond,
  symbolSquare,
  symbolStar,
  symbolTriangle,
  symbolWye,
} from 'd3-shape';

import { ChartProps } from '../../../types';
import { transition } from 'd3-transition';

export interface LollipopHChartProps extends ChartProps {
  classNamePoints?: string;
  classNameLines?: string;
  classNameSymbols?: string;
  tooltip?: {
    className?: string;
    html?: (d: any) => string;
    keys?: string[];
  };
  shape?:
    | 'circle'
    | 'diamond'
    | 'triangle'
    | 'square'
    | 'cross'
    | 'star'
    | 'wye';
  x?: {
    key: string;
    start?: number;
    end?: number;
    axis?: 'top' | 'bottom';
    axisTicks?: number;
  };
  y?: {
    key: string;
    axis?: 'left' | 'right';
  };
}

const LollipopHChart = ({
  data = [],
  id,
  className,
  classNamePoints,
  classNameLines,
  classNameSymbols,
  margin = {
    left: 80,
    right: 40,
    top: 40,
    bottom: 40,
  },
  padding = {
    left: 10,
    right: 0,
    top: 0,
    bottom: 20,
  },
  tooltip = undefined,
  shape = 'circle',
  x = { key: 'x', axis: 'bottom', axisTicks: 2 },
  y = { key: 'y', axis: 'left' },
}: LollipopHChartProps) => {
  const refreshChart = React.useCallback(() => {
    const svg = select(`#${id}`);

    // Clear svg

    svg.selectAll('*').remove();

    const width = +svg.style('width').split('px')[0],
      height = +svg.style('height').split('px')[0];

    data.sort((a: any, b: any) => b[x.key] - a[x.key]);

    const shapeMapping = {
      circle: symbolCircle,
      diamond: symbolDiamond,
      triangle: symbolTriangle,
      square: symbolSquare,
      cross: symbolCross,
      star: symbolStar,
      wye: symbolWye,
    };

    const xFn = scaleLinear()
      .domain([
        Number.isFinite(x.start) ? x.start : min(data, (d: any) => d[x.key]),
        Number.isFinite(x.end) ? x.end : max(data, (d: any) => d[x.key]),
      ])
      .range([
        (margin.left || 0) + (padding.left || 0),
        width - (padding.right || 0) - (margin.right || 0),
      ]);

    const yFn = scalePoint()
      .domain(data.map((d: any) => d[y.key]))
      // .range([height + (margin.top ||0) - (padding.bottom ||0), (margin.top ||0) + (padding.top ||0)])
      .range([
        (margin.top || 0) + (padding.top || 0),
        height - (margin.bottom || 0) - (padding.bottom || 0),
      ]);

    const g = svg.append('g');

    const xAxis =
      x.axis === 'top'
        ? axisTop(xFn).ticks(x.axisTicks || 2)
        : axisBottom(xFn).ticks(x.axisTicks || 2);

    const xAxisG = g.append('g').attr('class', 'axis--x axis ');

    xAxisG
      .attr(
        'transform',
        `translate(0, ${
          x.axis === 'top' ? margin.top || 0 : height - (margin.bottom || 0)
        })`
      )
      .call(xAxis);

    const yAxis = y.axis === 'right' ? axisRight(yFn) : axisLeft(yFn);

    const yAxisG = g
      .append('g')
      .attr('class', 'yAxis axis')
      .attr(
        'transform',
        `translate(${
          y.axis === 'right' ? (margin.left || 0) + width : margin.left || 0
        },0)`
      );

    (padding.left || 0) &&
      xAxisG
        .append('line')
        .attr('x1', margin.left || 0)
        .attr('x2', (margin.left || 0) + (padding.left || 0))
        .attr('y1', 0)
        .attr('y2', 0)
        .attr('stroke', 'currentColor');

    (padding.bottom || 0) &&
      yAxisG
        .append('line')
        .attr('x1', 0)
        .attr('x2', 0)
        .attr('y1', (margin.top || 0) + height - (padding.bottom || 0))
        .attr('y2', (margin.top || 0) + height)
        .attr('stroke', 'currentColor');

    yAxisG.call(yAxis);

    const dataGroup = g.append('g');

    const drawLinesAndCircles = () => {
      const pointGroup = dataGroup
        .selectAll('.line')
        .data(data)
        .enter()
        .append('g')
        .on('mouseover', function (event, d: any) {
          tooltip && tooltipDiv.style('opacity', 1);
          const [bX, bY] = pointer(event, select('body'));
          tooltipDiv.style('left', `${bX + 10}px`).style('top', `${bY + 10}px`);
          tooltipDiv.html(
            tooltip && tooltip.html
              ? tooltip.html(d)
              : tooltip?.keys
              ? tooltip?.keys
                  .map((key) => `${key}: ${d[key] || ''}`)
                  .join('<br/>')
              : `${d[y.key]}: ${d[x.key]}`
          );
        })
        .on('mouseleave', function () {
          tooltip && tooltipDiv.style('opacity', '0');
        });

      transition();

      pointGroup
        .append('line')
        .attr(
          'class',
          twMerge(
            `stroke-2 stroke-current`,
            classNamePoints || '',
            classNameLines || ''
          )
        )
        .attr('x1', (margin.left || 0) + (padding.left || 0))
        .attr('y1', (d: any) => yFn(d[y.key]) || 0)
        .attr('x2', margin.left || 0)
        .attr('y2', (d: any) => yFn(d[y.key]) || 0)
        .transition()
        .duration(1000)
        .attr('x2', (d: any) => xFn(d[x.key]));

      pointGroup
        .append('path')
        .attr('class', twMerge(classNamePoints || '', classNameSymbols || ``))
        .attr('d', () => symbol(shapeMapping[shape], 100)())
        .attr(
          'transform',
          (d: any) =>
            `translate(${(margin.left || 0) + (padding.left || 0)},${yFn(
              d[y.key]
            )})`
        )
        .transition()
        .duration(1000)
        .attr(
          'transform',
          (d: any) => `translate(${xFn(d[x.key])},${yFn(d[y.key])} )`
        );
    };

    drawLinesAndCircles();

    const tooltipDiv = select('body')
      .append('div')
      .attr('id', `tooltip-${id}`)
      .style('position', 'absolute')
      .style('opacity', '0')
      .attr('class', twMerge(`tooltip`, tooltip?.className || ''));
  }, [
    classNameLines,
    classNamePoints,
    classNameSymbols,
    data,
    id,
    margin,
    padding,
    shape,
    tooltip,
    x,
    y,
  ]);

  useEffect(() => {
    refreshChart();
    return () => {
      selectAll(`#tooltip-${id}`).remove();
    };
  }, [data, refreshChart]);
  return (
    <svg
      id={id}
      className={twMerge(defaultChartClassNames, className)}
      data-testid='bar-chart'
    />
  );
};

export default LollipopHChart;
