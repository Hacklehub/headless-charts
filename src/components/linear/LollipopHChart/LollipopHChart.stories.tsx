/* eslint-disable @typescript-eslint/no-explicit-any */
import { Meta, StoryObj } from '@storybook/react';

import LollipopHChart from '.';
import data from './sample.json';

export default {
  title: 'Linear/LollipopHChart/Intro',
  component: LollipopHChart,
  tags: ['autodocs'],
} as Meta;

type Story = StoryObj<typeof LollipopHChart>;

export const Default: Story = {
  args: {
    data,
    id: 'lollipop-h-chart',
    x: {
      key: 'value',
      start: 0,
    },
    y: {
      key: 'name',
    },
  },
};

export const WithCustomShape: Story = {
  args: {
    ...Default.args,
    shape: 'diamond',
  },
};
