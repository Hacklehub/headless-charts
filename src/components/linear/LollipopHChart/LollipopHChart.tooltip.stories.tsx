/* eslint-disable @typescript-eslint/no-explicit-any */
import { Meta, StoryObj } from '@storybook/react';

import LollipopHChart from '.';
import data from './sample.json';

const meta: Meta<typeof LollipopHChart> = {
  title: 'Linear/LollipopHChart/Tooltips',
  component: LollipopHChart,
  tags: ['autodocs'],
};

export default meta;

type Story = StoryObj<typeof LollipopHChart>;

export const Tooltip: Story = {
  args: {
    data,
    id: 'tooltip',
    x: {
      key: 'value',
      start: 0,
    },
    y: {
      key: 'name',
    },
    tooltip: {
      keys: ['name', 'value', 'reading'],
    },
  },
};

export const TooltipCustomHtml: Story = {
  args: {
    ...Tooltip.args,
    tooltip: {
      html: (d: any) =>
        `<div class='bg-gray-800 text-white p-2 rounded'>${d.name} - ${d.value}</div>`,
    },
  },
};
