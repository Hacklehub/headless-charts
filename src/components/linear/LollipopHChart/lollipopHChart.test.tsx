import { describe, expect, it } from 'vitest';
import { render, screen } from '@testing-library/react';

import LollipopHChart from '.';
import data from './sample.json';

describe('LollipopHChart', () => {
  it('renders lollipopHChart component for snapshot test', () => {
    render(<LollipopHChart id='lollipopHChart' data={data} />);
    const lollipopHChart = screen.getByTestId('lollipopHChart');
    expect(lollipopHChart).toMatchSnapshot();
  });
  it('renders lollipopHChart component', () => {
    render(<LollipopHChart id='lollipopHChart' data={data} />);
    const lollipopHChart = screen.getByTestId('lollipopHChart');
    expect(lollipopHChart).toBeInTheDocument();
  });
});
