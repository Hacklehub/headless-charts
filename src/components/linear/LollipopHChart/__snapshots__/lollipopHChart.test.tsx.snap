// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LollipopHChart > renders lollipopHChart component for snapshot test 1`] = `
<svg
  class="chart dark:bg-gray-800 dark:text-gray-50 h-64 lg:w-4/12 md:w-6/12 text-gray-900 w-full"
  data-testid="lollipopHChart"
  id="lollipop-h-chart"
>
  <g>
    <g
      class="axis--x axis "
      fill="none"
      font-family="sans-serif"
      font-size="10"
      text-anchor="middle"
      transform="translate(0, -40)"
    >
      <path
        class="domain"
        d="M90.5,6V0.5H-39.5V6"
        stroke="currentColor"
      />
      <g
        class="tick"
        opacity="1"
        transform="translate(90.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          0
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(25.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          10
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(-39.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          20
        </text>
      </g>
      <line
        stroke="currentColor"
        x1="80"
        x2="90"
        y1="0"
        y2="0"
      />
    </g>
    <g
      class="yAxis axis"
      fill="none"
      font-family="sans-serif"
      font-size="10"
      text-anchor="end"
      transform="translate(80,0)"
    >
      <line
        stroke="currentColor"
        x1="0"
        x2="0"
        y1="20"
        y2="40"
      />
      <path
        class="domain"
        d="M-6,40.5H0.5V-59.5H-6"
        stroke="currentColor"
      />
      <g
        class="tick"
        opacity="1"
        transform="translate(0,40.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          Team B
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,-59.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          Team A
        </text>
      </g>
    </g>
    <g>
      <g>
        <line
          class="stroke-current"
          x1="90"
          x2="80"
          y1="40"
          y2="40"
        />
        <path
          class=""
          d="M5.642,0A5.642,5.642,0,1,1,-5.642,0A5.642,5.642,0,1,1,5.642,0"
          transform="translate(90,40)"
        />
      </g>
      <g>
        <line
          class="stroke-current"
          x1="90"
          x2="80"
          y1="-60"
          y2="-60"
        />
        <path
          class=""
          d="M5.642,0A5.642,5.642,0,1,1,-5.642,0A5.642,5.642,0,1,1,5.642,0"
          transform="translate(90,-60)"
        />
      </g>
    </g>
  </g>
</svg>
`;
